src/Database/Type/EnumType.php

```diff
@@ -0,0 +1,185 @@
+use Cake\Utility\Text;
+use InvalidArgumentException;
+use PDO;
+use ReflectionEnum;
+use ReflectionException;
+
+/**
+ * Enum type converter.
+ *
+ * Use to convert string data between PHP and the database types.
+ */
+class EnumType extends BaseType
+{
+    /**
+     * The type of the enum which is either string or int
+     *
+     * @var string
+     */
+    protected string $backingType;
+
+    /**
+     * @param string $name The name identifying this type
+     * @param string $enumClassName The associated enum class name
+     */
+    public function __construct(
+        string $name,
+        protected string $enumClassName
+    ) {
+        parent::__construct($name);
+        try {
+            $reflectionEnum = new ReflectionEnum($enumClassName);
+            $this->backingType = (string)$reflectionEnum->getBackingType();
+        } catch (ReflectionException) {
+            throw new DatabaseException(
+                sprintf('Given enum %s is not a backed enum', $enumClassName)
+            );
+        }
+    }
+
+    /**
+     * Convert enum instances into the database format.
+     *
+     * @param mixed $value The value to convert.
+     * @param \Cake\Database\Driver $driver The driver instance to convert with.
+     * @return string|int|null
+     */
+    public function toDatabase(mixed $value, Driver $driver): string|int|null
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        if ($value instanceof BackedEnum) {
+            return $value->value;
+        }
+
+        if (is_int($value) || is_string($value)) {
+            return $value;
+        }
+
+        throw new InvalidArgumentException(sprintf(
+            'Cannot convert value of type `%s` to string or integer',
+            get_debug_type($value)
+        ));
+    }
+
+    /**
+     * Transform DB value to backed enum instance
+     *
+     * @param mixed $value The value to convert.
+     * @param \Cake\Database\Driver $driver The driver instance to convert with.
+     * @return \BackedEnum|null
+     */
+    public function toPHP(mixed $value, Driver $driver): ?BackedEnum
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        if (get_debug_type($value) !== $this->backingType) {
+            throw new InvalidArgumentException(sprintf(
+                'Given value type `%s` does not match associated `%s` backed enum',
+                get_debug_type($value),
+                $this->backingType
+            ));
+        }
+
+        return $this->enumClassName::tryFrom($value);
+    }
+
+    /**
+     * Get the correct PDO binding type for string or integer data.
+     *
+     * @param mixed $value The value being bound.
+     * @param \Cake\Database\Driver $driver The driver.
+     * @return int
+     */
+    public function toStatement(mixed $value, Driver $driver): int
+    {
+        if (is_int($value)) {
+            return PDO::PARAM_INT;
+        }
+
+        return PDO::PARAM_STR;
```

Comment:
Should this use the `backingType`?

Reply:
If I do so then `$this->type->toStatement(1, $this->driver)` would return `PDO::PARAM_STR` because my test currently only uses a string backed enum.

Reply:
I don't follow. Are you saying this code is to help a test pass?

If a use has an string enum and passes an int, is that valid?

Reply:
After reading a bit more into the whole topic I see that `toDatabase()` is called before `toStatement()` inside the `TypeConverterTrait`. 

Therefore we should add another check to the `toDatabase()` method to prevent users form passing a valid int backed enum to a field which has a string backed enum connected. Currently we are only getting the value of the given enum instance, no matter if its actually the associated enum type or not.

Reply:
I have now adjusted the `toStatement()` method to use the `backingType` as described above because we now check if the given backed enum value is actually an instance of the connected enumClass in the current type instance.